<a href="https://colab.research.google.com/github/csanicola74/Important_Reference_Repo/blob/main/w3resource_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Learning Python through W3Resource.com
Links:
  * [Python](https://www.w3resource.com/python/python-syntax.php)
  * [Exercises](https://www.w3resource.com/python-exercises/)

# Exercise

## #1
Write a Python program to print the following string in a specific format (see the output).

In [27]:
# Sample String: "Twinkle, twinkle, little star, How I wonder what you are! Up above the world so high, Like a diamond in the sky. Twinkle, twinkle, little star, How I wonder what you are"
# Output:
  # Twinkle, twinkle, little star,
	    # How I wonder what you are!
	            #	Up above the world so high,
	            #	Like a diamond in the sky.
  # Twinkle, twinkle, little star,
	    # How I wonder what you are
#################
### REFERENCE ###
#################
# \n_________ = new line
# \n\t_______ = new line and a tab in
# \n\t\t_____ = new line and 2 tabs in
print("Twinkle, twinkle, little star, \n\tHow I wonder what you are! \n\t\tUp above the world so high, \n\t\tLike a diamond in the sky. \nTwinkle, twinkle, little star, \n\tHow I wonder what you are!")

Twinkle, twinkle, little star, 
	How I wonder what you are! 
		Up above the world so high, 
		Like a diamond in the sky. 
Twinkle, twinkle, little star, 
	How I wonder what you are!


The print statement (Python 2.6) has been replaced with a print() function(Python 2.6), with keyword arguments to replace most of the special syntax of  the old print statement .
The print() function doesn’t support the “softspace” feature of the old print statement. For example, in Python 2.x, print "A\n", "B" would write "A\nB\n"; but in Python 3.0, print("A\n", "B") writes "A\n B\n".
Initially, you’ll be finding yourself typing the old print x a lot in interactive mode. Time to retrain your fingers to type print(x) instead!
When using the 2to3 source-to-source conversion tool, all print statements are automatically converted to print() function calls, so this is mostly a non-issue for larger projects.

## #2
Write a Python program to find out what version fo Python you are using.

In [28]:
# version 1
import sys  # Import the sys module to access system-specific parameters and functions

# Print the Python version to the console
print("Python version")

# Use the sys.version attribute to get the Python version and print it
print(sys.version)

# Print information about the Python version
print("Version info.")

# Use the sys.version_info attribute to get detailed version information and print it
print(sys.version_info)


# version 2
import platform
print(platform.python_version())

Python version
3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
Version info.
sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
3.10.12


**Explanation:**
The said code imports the "sys" module and then uses it to print the version of Python currently being used, as well as additional version information.
The first line imports the "sys" module provides access to some variables used or maintained by the interpreter and to functions that interact strongly with the interpreter. It is always available.
The second line prints the string "Python version".
The third line prints the version of Python currently being used to the console.
The fourth line prints the string "Version info.".
The fifth line prints version information of the current Python interpreter. It returns a named tuple (version_info) containing the five components of the version number: major, minor, micro, releaselevel, and serial.


## #3
Write a Python program to display the current date and time.

In [29]:
# Sample Output:
# Current date and time :
# 2014-07-05 14:34:14

# Import the 'datetime' module to work with date and time
import datetime

# Get the current date and time
now = datetime.datetime.now()

# Create a datetime object representing the current date and time

# Display a message indicating what is being printed
print("Current date and time: ")

# Print the current date and time in a specific format
print(now.strftime("%Y-%m-%d %H:%M:%S"))

# Use the 'strftime' method to format the datetime object as a string with the desired format

Current date and time: 
2024-03-20 19:56:29


**Explanation:**
The said code imports the "datetime" module, gets the current date and time, and finally prints it in a formatted string.
The first line import datetime imports the datetime module which supplies classes for manipulating dates and times.
The second line now = datetime.datetime.now() creates a datetime object for the current date and time.
The third line print ("Current date and time : ") prints the string "Current date and time : " to the console.
The fourth line print (now.strftime("%Y-%m-%d %H:%M:%S")) uses the strftime() method of the datetime object to format the date and time as a string in the format "YYYY-MM-DD HH:MM:SS" and prints it to the console.
The strftime() method returns a string representing the date, controlled by an explicit format string. The format codes used in this example are:
%Y: year with century as a decimal number.
%m: month as a zero-padded decimal number.
%d: day of the month as a zero-padded decimal number.
%H: hour (24-hour clock) as a zero-padded decimal number.
%M: minute as a zero-padded decimal number.
%S: second as a zero-padded decimal number.
This code can be useful for logging or timestamping events in a program, or for displaying the current date and time in a user interface.

## #4
Write a Python program that calculates the area of a circle based on the radius entered by the user.

In [30]:
# Sample Output:
# r = 1.1
# Area = 3.8013271108436504

# Import the 'pi' constant from the 'math' module to calculate the are of a circle
from math import pi

# Prompt the user to input the radius of the circle
r = float(input("Input the radius of the circle: "))

# Calculate the are of the circle using the formula: area = π * r^2
area = pi * r ** 2

# Dispaly the result, including the radius and calculated area
print("The area of the circle with radius " + str(r) + " is: " + str(area))

Input the radius of the circle: 12
The area of the circle with radius 12.0 is: 452.3893421169302


**Explanation:**
The said code calculates the area of a circle based on the radius entered by the user. The code uses the "math" module's pi constant and the "input" function to get the radius from the user, then it uses the formula to calculate the area of the circle.
The first line from math import pi imports the pi constant from the math module, which is the mathematical constant. It represents the ratio of a circle's circumference to its diameter (approximately equal to 3.14).
The second line r = float(input ("Input the radius of the circle : ")) gets the radius of the circle from the user using the input() function. input() function assigns it to the variable r, it's then cast to float, so the user can input decimal number also.
The third line print ("The area of the circle with radius " + str(r) + " is: " + str(pi * r**2)) uses the formula to calculate the area of the circle (pi*r**2) . Then it concatenates the string and the value of the radius and area using the + operator and prints the final string.
This code is useful to calculate the area of a circle when the radius is known, it can be used in physics and mathematical calculations or in applications that require the area of a circle to be calculated.

## #5
Write a Python program that accepts the user's first and last name and prints them in reverse order with a space between them

In [31]:
first_name = input("What is your first name: ")
last_name = input("What is your last name: ")
print(last_name + " " + first_name)

What is your first name: Caroline
What is your last name: Sanicola
Sanicola Caroline


**Explanation:**
This code prompts the user to input their first and last name, and then it prints a message that combines the first name and last name in a specific order.
The first line fname = input("Input your First Name : ") gets the user's first name using the input() function and assigns it to the variable 'fname'.
The second line lname = input("Input your Last Name : ") gets the user's last name using the input() function and assigns it to the variable 'lname'.
The third line print ("Hello " + lname + " " + fname) prints a message that combines the last name and first name, concatenating them with the + operator and separated by a space.

## #6
Write a Python program that accepts a sequence of comma-separated numbers from the user and generates a list of a tuple from those numbers.

In [32]:
# Sample data: 3, 5, 7, 23
# Output:
#   List: ['3', '5', '7', '23']
#   Tuple: ('3', '5', '7', '23')

# Prompt the user to input a sequence of comma-separated numbers and store it in the 'values' variable
values = input("Input some comma-separated numbers: ")

# Split the 'values' string into a list using commas as separators and store it in the 'list' variable
list = values.split(",")

# Convert the 'list' into a tuple and store it in the 'tuple' variable
tuple = tuple(list)

# Print the list
print('List: ', list)

# Print the tuple
print('Tuble: ', tuple)

Input some comma-separated numbers: 4, 5, 23, 87, 22, 314
List:  ['4', ' 5', ' 23', ' 87', ' 22', ' 314']
Tuble:  ('4', ' 5', ' 23', ' 87', ' 22', ' 314')


**Explanation:**
The said code prompts the user to input a list of numbers separated by commas. Then it converts the string of numbers into a list and a tuple, and finally it prints both the list and the tuple.
The first line values = input("Input some comma seprated numbers : ") gets a string of numbers separated by commas from the user using the input() function and assigns it to the variable 'values'.
The second line splits the string of numbers into a list of individual numbers using the split() method and assigns it to the variable 'list'. This method takes a separator as input, in this case ',' (comma) and returns a list of substrings that were separated by the separator.
The third line converts the list into a tuple using the tuple() function and assigns it to the variable 'tuple'.
The fourth line prints the list to the console.
The fifth line prints the tuple to the console.

## #7
Write a Python program that accepts a filename from the user and prints the extension of the file.

In [33]:
# Sample filename: abc.java
# Output: java

# Prompt the user to input a filename and store it in the 'filename' variable
filename = input("Input the Filename: ")

# Split the 'filename' string into a list using the period (.) as a separator and store it in the 'f_extns' variable
f_extns = filename.split(".")

# Print the extension of the file, which is the last element in the 'f_extns' list
print("The extension of the file is: " + repr(f_extns[-1]))

Input the Filename: data.xlsx
The extension of the file is: 'xlsx'


**Explanation:**
At first the siad code prompts the user to input a filename and stores it in the variable "filename". It then uses the string method split() to split the value of "filename" by "."(dot) and assigns the result to the variable "f_extns".
The code then uses the index -1 to access the last element of the list "f_extns" which is the file extension and prints the message "The extension of the file is : " followed by the last element.
The repr() function returns a string containing a printable representation of an object.
For example, if the user inputs "abc.txt" as the filename, the code will print "Th e extension of the file is : 'txt'"

## #8
Write a Python program to display the first and last colors from the follow list.

In [34]:

# color_list = ["Red", "Green", "White", "Black"]

# Create a list called 'color_list' containing color names
color_list = ["Red", "Green", "White", "Black"]
# Print the first and last elements of the 'color_list' using string formatting
# The '%s' placeholders are filled witht he values of 'color_list[0]' (Red) and 'color_list[-1]' (Black)
print("%s %s" % (color_list[0], color_list[-1]))

Red Black


**Explanation:**
The said code creates a list named "color_list" which includes four different colors, "Red", "Green", "White", and "Black". Then, it utilizes string formatting to print the first (Index 0) and the last (Index -1) color of the list together with a space in between. The output would be "Red Black".

## #9
Write a Python program to display the examination schedule. (extract the date from exam_st_date).

In [35]:
# exam_st_date = (11, 12, 2014)
# Sample Output: The examination will start from: 11/12/2014

# Define a tuple called 'exam_st_date' containing the exam start date in the format (day, month, year)
exam_st_date = (11, 12, 2014)

# Print the exam start date using string formatting
# The '%i' placeholders are filled with the values from the 'exam_st_date' tuple
print("The examination will start from: %i/%i/%i" % exam_st_date)

The examination will start from: 11/12/2014


**Explanation:**
The said code creates a tuple called "exam_st_date" containing three integers: 11, 12, and 2014. It then uses string formatting to print a string that states "The examination will start from :" followed by the integers in the tuple in the order they appear, formatted as day/month/year. The placeholders %i are used to format the integers.
The output would be: "The examination will start from : 11 / 12 / 2014".

## #10
Write a Python program that accepts an integer (n) and computes the value of n+nn+nnn


In [36]:
# Sample value of n is 5
# Expected Result: 615

# Prompt the user to input an integer and store it in the variable 'a'
a = int(input("Input an integer: "))

# Create new integers 'n1', 'n2', and 'n3' by concatenating 'a' with itself one, two, and three times, respectively
n1 = int("%s" % a)              # Convert 'a' to an integer
n2 = int("%s%s" % (a, a))       # Concatenate 'a' with iteself and convert to an integer
n3 = int("%s%s%s" % (a, a, a))  # Concatenate 'a' with itself twice and convert to an integer

# Calculate the sum of 'n1', 'n2', and 'n3' and print the result
print(n1 + n2 + n3)

Input an integer: 63
642789


**Explanation:**
The said code prompts the user to input an integer, which is then stored in the variable "a".
There after the first variable n1 is created by converting the input integer "a" to a string and then back to an integer.
The second variable n2 is created by concatenating two copies of the input integer "a" as a string and then converting that string to an integer.
The third variable n3 is created by concatenating three copies of the input integer "a" as a string and then converting that string to an integer.
Finally, the code sums the three variables n1, n2 and n3 and print the result.

## #11
Write a Python program to print the documents (syntax, description etc.) of Python built-in function(s).

In [37]:
# Sample function: abs()
# Expected Result:
# abs(number) -> number
# Return the absolute value of the argument

# Print the docstring (documentation) of the 'abs' function
print(abs.__doc__)

Return the absolute value of the argument.


Python Docstring:
A docstring is a string literal that occurs as the first statement in a module, function, class, or method definition. Such a docstring becomes the __doc__ special attribute of that object.
All modules should normally have docstrings, and all functions and classes exported by a module should also have docstrings. Public methods (including the __init__ constructor) should also have docstrings.

## #12
Write a Python program that prints the calendar for a given month and year.

In [38]:
# Note: Use 'calendar' module

# Import the 'calendar' module
import calendar

# Prompt the user to input the year and month
y = int(input("Input the year: "))
m = int(input("Input the month: "))

# Print the calendar for the specified year and month
print(calendar.month(y, m))

Input the year: 2024
Input the month: 3
     March 2024
Mo Tu We Th Fr Sa Su
             1  2  3
 4  5  6  7  8  9 10
11 12 13 14 15 16 17
18 19 20 21 22 23 24
25 26 27 28 29 30 31



**Explanation:**
The said code imports the calendar module, which allows you to output calendars like the Unix cal program, and provides additional useful functions related to the calendar. It then asks the user to input an integer for the year and month.
The code then uses the month() function from the calendar module, which takes the year and month as arguments and returns a string containing an ASCII calendar for that month of that year.

## #13
Write a Python program to print the following 'here document'.

In [39]:
# Sample string:
# a string that you "don't" have to escape
# This
# is a ..... multi-line
# heredoc string ------> example

print("""
a string that you "don't" have to escape
This
is a ....... multi-line
heredoc string --------> example
""")


a string that you "don't" have to escape
This
is a ....... multi-line
heredoc string --------> example



**Explanation:**
A here document (or "heredoc") is a way of specifying a text block, preserving the line breaks, indents, and other whitespace within the text. The said code uses triple quotes to define a multi-line string, also known as a heredoc string. It allows the string to span multiple lines without the need to escape newline characters.
The output will be the string exactly as it appears in the code, including the newline characters.

## #14
Write a Python program to calculate the number of days between two dates.

In [40]:
# Sample dates: (2014, 7, 2), (2014, 7, 11)
# Expected output: 9 days

# Import the 'date' class from the 'datetime' module
from datetime import date

# Define a start date as July 2, 2014
f_date = date(2014, 7, 2)

# Define an end date as July 11, 2014
l_date = date(2014, 7, 11)

# Calculate the difference between the end date and start date
delta = l_date - f_date

# Print the number of days in the time difference
print(delta.days)

9


**Explanation:**
This script imports the "date" class from the "datetime" module.
It then creates two date objects (<class 'datetime.date'>), 'f_date 'and 'l_date'.
The script then calculates the difference between these two dates and stores it in a variable called 'delta'.
The 'delta' variable is then printed, specifically the attribute "days" which returns the number of days between the said two dates.
In this case, the output would be 9, as there are 9 days between July 2nd and July 11th in 2014.

## #15
Write a Python program to get the volume of a sphere with radius six

In [41]:
# Define the value of pi
pi = 3.1415926535897931

# Define the radius of the sphere
r = 6.0

# Calculate the volume of the sphere using the formula
V = 4.0/3.0 * pi * r**3

# Print the calculated volume of the sphere
print('The volumne of the sphere is: ', V)

The volumne of the sphere is:  904.7786842338603


**Explanation:**
The said code assigns the value of pi as 3.1415926535897931 to the variable "pi" and 6.0 to the variable "r". The code uses the formula V = 4/3 * pi * r^3 to calculate the volume of a sphere. Volume of the sphere is stored in the variable "V".
It then prints the message "The volume of the sphere is:" followed by the volume of the sphere using print statement.
In this case, the code will output "The volume of the sphere is: 904.7786842338603".

## #16
Write a Python program to calculate the difference between a given number and 17.

In [42]:
# If the number is greater than 17, return twice the absolute difference.

# Define a function named "difference" that takes an integer parameter "n"
def difference(n):
  # Check if n is less than or equal to 17
  if n <= 17:
    # If n is less than or equal to 17, return the absolute difference between 17 and n
    return 17 - n
  else:
    # If n is greater than 17, return the absolute difference between n and 17 multiplied by 2
    return (n - 17) * 2

# Call the "difference" function with the argument 22 and print the result
print(difference(22))

# Call the "difference" function with the argument 14 and print the result
print(difference(14))

10
3


**Explanation:**
The said script defines a function called 'difference' that takes input(n) as an argument. Using an if-else statement the function checks whether the input value of n is less than or equal to 17. If the input value of n is less than or equal to 17, the function will return the difference of 17 - n. If the input value of n is greater than 17, the function will return double the difference of n - 17.
Now the script calls the function twice with the input values of 22 and 14, respectively.
In the first case the input is greater than 17, so the function returns (22-17)*2 = 10. Therefore, the script will print 10.
In the second case the input value is less than or equal to 17, so the function returns 17-14 = 3. Therefore, the script will print 3.

## #17
Write a Python program to test whether a number is within 100 of 1000 or 2000.

In [43]:
# Define a function named "near_thousand" that takes an integer parameter "n"
def near_thousand(n):
  # Check if the absolute difference between 1000 and n is less than or equal to 100
  # OR check if the absoluate difference between 2000 and n is less than or equal to 100
  return ((abs(1000 - n) <= 100) or (abs(2000 - n) <= 100))

# Call the "near_thousand" function with the argument 1000 and print the result
print(near_thousand(1000))

# Call the "near_thousand" function with the argument 900 and print the result
print(near_thousand(900))

# Call the "near_thousand" function with the argument 800 and print the result
print(near_thousand(800))

# Call the "near_thousand" function with the argument 2200 and print the result
print(near_thousand(2200))

True
True
False
False


**Explanation:**
The said script defines a function called near_thousand(n) that takes an integer as an argument and returns True if the input is within 100 of either 1000 or 2000, and False otherwise. The function uses the abs() function, which returns the absolute value of a number, to determine the difference between the input and 1000 or 2000.
Now the script calls the function four time with the input values of 1000, 900, 800 and 2200 respectively.
When the input value is 1000, it returns True because the difference between 1000 and 1000 is 0, which is less than 100.
When the input value is t 900, it returns True because the difference between 900 and 1000 is 100, which is less than 100.
When the input value is 800, it returns False because the difference between 800 and 1000 is 200, which is not less than or equal to 100.
When the input value is 2200, it returns False because the difference between 2200 and 2000 is 200, which is not less than or equal to 100.

## #18
Write a Python program to calculate the sum of three given numbers.

In [44]:
# If the values are equal, return three times their sum.

# Define a function named "sum_thrice" that takes three integer parameters: x, y, and z
def sum_thrice(x, y, z):
  # Calculate the sum of x, y, and z
  sum = x + y + z

  # Check if x, y, and z are all equal (all three numbers are the same)
  if x == y == z:
    # If they are equal, triple the sum
    sum = sum * 3

  # Return the final sum
  return sum

# Call the "sum_thrice" function with the arguments (1, 2, 3) and print the result
print(sum_thrice(1, 2, 3))

# Call the "sum_thrice" function with the arguments (3, 3, 3) and print the result
print(sum_thrice(3, 3, 3))

6
27


**Explanation:**
The said script defines a function called sum_thrice(x, y, z) that takes three integers as arguments and returns their sum. The function, however, multiplies the sum by 3 if all three integers are equal.
Now the script calls the function twice with the input values (1,2,3) and (3,3,3).
Whenever 1, 2, and 3 are input values, the three inputs are not equal, so it returns 6.
Whenever 3, 3, and 3 are input values, the three inputs are equal so the function multiplies the sum by 3 and returns 27.
Depending on the inputs, the if statement holds true only if x == y == z, otherwise it holds False.

## #19
Write a Python program to get a newly-generated string from a given string where "ls" has been added to the front.

In [45]:
# Return the string unchanged if the given string already begins with "ls"

# Define a function named "new_string" that takes a string parameter called "text"
def new_string(text):
  # Check if the length of the "text" is greater than or equal to 2 and if the first two characters of "text" are "Is"
  if len(text) >= 2 and text[:2] == "Is":
    # If the conditions are met, return the original "text" unchanged
    return text
  else:
    # If the conditions are not met, prepend "Is" to the "text" and return the modified string
    return "Is" + text

# Call the "new_string" function with the argument "Array" and print the result
print(new_string("Array"))

# Call the "new_string" function with the argument "IsEmpty" and print the result
print(new_string("IsEmpty"))

IsArray
IsEmpty


**Explanation:**
The said script defines a function "new_string(text)" takes a string as an argument, and checks whether the string's length is greater than or equal to 2 and the first two characters of the string are "Is". If both conditions are true, the function returns the original string. Otherwise, the function returns a new string "Is" concatenated with the original string.
The first print statement will display "IsArray" and the second will display "IsEmpty".

## #20
Write a Python program that returns a string that is n (non-negative integer) copies of a given string.

In [46]:
# Define a function named "larger_string" that takes two parameters, "text" and "n"
def larger_string(text, n):
    # Initialize an empty string variable named "result"
    result = ""

    # Use a for loop to repeat the "text" "n" times and concatenate it to the "result"
    for i in range(n):
        result = result + text

    # Return the final "result" string
    return result

# Call the "larger_string" function with the arguments 'abc' and 2, then print the result
print(larger_string('abc', 2))

# Call the "larger_string" function with the arguments '.py' and 3, then print the result
print(larger_string('.py', 3))

abcabc
.py.py.py


**Explanation:**
The said code defines a function called "larger_string" which takes two arguments: "text" and "n". The function will concatenate the "text" argument "n" number of times and return the result as a new string.
The function starts by initializing an empty string variable called "result". Then, it uses a for loop to iterate over a range of numbers, starting from 0 and ending at "n". On each iteration, the function takes the current value of "result" and adds the "text" argument to it, effectively concatenating the "text" argument with itself "n" times.
The last two lines of code call the function, with different arguments.
In the first case the function takes 'abc' as text and 2 as the n, the function will concatenate 'abc' two times and return 'abcabc'
In the second case the function takes '.py' as text and 3 as the n, the function will concatenate '.py' three times and return '.py.py.py'
The function returns the concatenated string as the final output.

## #21
Write a Python program that determines whether a given number (accepted from the user) is even or odd, and prints an appropriate message to the user

In [47]:
# Prompt the user to enter a number and convert the input to an integer
num = int(input("Enter a number: "))

# Calculate the remainder when the number is divided by 2
mod = num % 2

# Check if the remainder is greater than 0, indicating an odd number
if mod > 0:
  # Print a message indicating that the number is odd
  print("this is an odd number.")
else:
  # Print a message indicating that the number is even
  print("this is an even number.")

Enter a number: 6
this is an even number.


**Explanation:**
The said code prompts the user to input a number, then converts the input to an integer and assigns it to the variable 'num'. Then it calculates the remainder of 'num' and 2 and assigns it to the variable 'mod'. Next, it checks the value of 'mod'. If the value of 'mod' is greater than 0, it prints "This is an odd number." otherwise prints "This is an even number.".

## #22
Write a Python program to count the number 4 in a given list.

In [48]:
# Define a function called list_count_4 that takes a list of numbers (nums) as a parameter.
def list_count_4(nums):
  # Initialize a variable count to keep track of the count of occurrences of the number 4.
  count = 0

  # Iterate through each element (num) in the input list (nums).
  for num in nums:
    # Check if the current element (num) is equal to 4.
    if num == 4:
      # If the element is 4, increment the count by 1.
      count = count + 1

  # Return the final count after iterating through the list.
  return count

# Call the list_count_4 function with two different input lists and print the results.
print(list_count_4([1, 4, 6, 7, 4]))  # Output: 2 (There are two occurrences of 4 in the list.)
print(list_count_4([1, 4, 6, 4, 7, 4]))  # Output: 3 (There are three occurrences of 4 in the list.)

2
3


**Explanation:**
The said code defines a function called "list_count_4" that takes a list of numbers as its argument, "nums". Inside the function, it initializes a variable "count" to zero. Next, it uses a for loop to iterate through each element in the "nums" list. For each iteration, it checks if the current element i.e. num is equal to 4, if true it increments the "count" by 1. After the for loop completes, the function returns the final "count" of how many times the number 4 appears in the list.
The last two lines of code call the function, with the input values ([1, 4, 6, 7, 4]) and ([1, 4, 6, 4, 7, 4]) and print the results.
The first call will return 2, and the second call will return 3.

## #23
Write a Python program to get n (non-negative integer) copies of the first 2 characters of a given string.

In [49]:
# Return n copies of the whole string if the length is less than 2.

# Define a function called substring_copy that takes two parameters: text (a string) and n (an integer).
def substring_copy(text, n):
  # Set the initial value of flen (substring length) to 2.
  flen = 2

  # Check if flen is greater than the length of the input text.
  if flen > len(text):
    # If flen is greater, set it to the length of the text to avoid going out of bounds.
    flen = len(text)

  # Extract a substring of length flen from the beginning of the text.
  substr = text[:flen]

  # Initialize an empty string result to store the concatenated substrings.
  result = ""

  # Iterate n times to concatenate the substring to the result.
  for i in range(n):
    result = result + substr

  #Return the final result after concatenating substrings.
  return result

# Call the substring_copy function with two different inputs and print the results
print(substring_copy('abcdef', 2)) # Output: "abab" (substring "ab" repeated 2 times)
print(substring_copy('p', 3)) # Output: "ppp" (substring "p" repeated 3 times)

abab
ppp


**Explanation:**
The said code defines a function called "substring_copy" that takes two arguments, a string "text" and an integer "n".
Inside the function, it creates a variable "flen" and assigns it the value of 2. Next, it checks whether the value of "flen" is greater than the length of the input string "text". If true it assigns the value of len(text) to "flen" otherwise it creates a variable "substr" and assigns it the value of the first two characters of the input string "text" using string slicing method.
Then it creates an empty string called "result" and uses a for loop to iterate "n" times. During each iteration it concatenates the "substr" to the "result" string. Finally, it returns the value of the "result" variable.
The last two lines of code call the function, with the input values ('abcdef', 2) and ('p', 3) and print the results. The first call will return 'abab' and the second will return 'ppp'.


## #24
Write a Python program to test whether a passed letter is a vowel or not.

In [50]:
# Define a function called is_vowel that takes one parameter: char (a character).
def is_vowel(char):
  # Define a string called all_vowels containing all lowercase vowel characters.
  all_vowels = 'aeiou'

  # Check if the input character (char) is present in the all_vowels string.
  return char in all_vowels

# Call the is_vowel function with two different characters and print the results
print(is_vowel('c')) # Output: False (character 'c' is not a vowel)
print(is_vowel('e')) # Output: True (character 'e' is a vowel)

False
True


**Explanation:**
The said code defines a function called "is_vowel" which takes in one parameter "char". First, the function declares a variable "all_vowels" containing all lowercase vowels. The function then uses the "in" operator to check whether the value of "char" parameter is present within the "all_vowels" string, and returns the result.
In the last two lines the code call the function "is_vowel" twice, with the input values 'c' and 'e' and print the results.
In the first case, False will be returned, while in the second case, True will be returned.

## #25
Write a Python program that checks whether a specified value is contained within a group of values.

In [51]:
# Define a function called is_group_member that takes two parameters: group_data (a list) and n (an integer)
def is_group_member(group_data, n):
  # Iterate through the elements (values) in the group_data list.
  for value in group_data:
    # Check if the current value is equal to the given integer, n.
    if n == value:
      return True # If found, return True.
  return False # If the loop completes and no match is found, return False.

# Call the is_group_member function with two different lists and integers, and print the results.
print(is_group_member([1, 5, 8, 3], 3)) # Output: True (3 is in the list)
print(is_group_member([5, 8, 3], -1)) # Output: False (-1 is not in the list)

True
False


**Explanation:**
The said code defines a function called "is_group_member" that takes two arguments "group_data" a list and an integer "n". First, the function declares a for loop to iterate through each element in the "group_data" list. When the current element (value) value matches the value of "n", it returns True, which means the input integer is present in the "group_data" list. The function returns False if the for loop fails to find a match, meaning the input integer does not exist in "group_data".
In the last two lines the code call the function "is_group_member" twice, with the input values ([1, 5, 8, 3], 3) and ([5, 8, 3], -1) and print the results.
In the first case, False will be returned, while in the second case, True will be returned.

## #26
Write a Python program to create a histogram from a given list of integers

In [52]:
# Define a function called histogram that takes a list of items as a parameter.
def histogram(items):
  # Iterate through the items in the list.
  for n in items:
    output = '' # Initialize an empty string called output.
    times = n   # Set the times variable to the value of n

    # Use a while loop to append '*' to the output string 'times' number of times
    while times > 0:
      output += '*'
      times = times - 1 # Decrement the times variable

    # Print the resulting output string.
    print(output)

# Call the histogram function with a list of numbers and print the histogram
histogram([2, 3, 6, 5])

**
***
******
*****


**Explanation:**
The function "histogram" takes a parameter "items ", a list of integers. To iterate through each item in the "items" list, the function uses a for loop. For each iteration, it creates a varaible "output", and assigns the current element (n) of the loop to the variable "times".
Then it uses a while loop which will run as long as the value of "times" is greater than 0. Inside the while loop, it concatenates an asterisk (*) character to the "output" variable and decrements the value of "times" by 1 at each iteration. After the while loop completes, it prints the final value of "output" variable.
The last line of code calls the "histogram" function with a list of integers ([2, 3, 6, 5]) as an argument, it will print a histogram of asterisks, where the number of asterisks corresponds to the value of the integers in the list.

## #27
Write a Python program that concatenates all elements in a list into a string and returns it

In [53]:
# Define a function called concatenate_list_data that takes a list as a parameter.
def concatenate_list_data(lst):
  result = '' # Initialize an empty string called result.

  # Iterate through the elements in the list.
  for element in lst:
    result += str(element) # Convert each element to a string and concatenate it to the result

  return result # Return the concatenated string.

# Call the concatenate_list_data function with a list of numbers and print the result.
print(concatenate_list_data([1, 5, 12, 2]))

15122


**Explanation:**
The said code defines a function called "concatenate_list_data" that takes a list as its argument. Inside the function, it creates an empty string variable called "result".
Then it uses a for loop to iterate through each element in the input list. For each iteration, it converts the current element to a string using the str() function and concatenates it to the "result" variable.
Upon completion of the for loop, the variable "result" is returned.
The last line of code calls the "concatenate_list_data" function with a list of integers ([1, 5, 12, 2]) as an argument, it will print a string "15122".

## #28
Write a Python program to print all even numbers from a given list of numbers in the same order and stop printing any after 237 in the sequence.

In [54]:
# Define a list of numbers.
numbers = [
    386, 462, 47, 418, 907, 344, 236, 375, 823, 566, 597, 978, 328, 615, 953, 345,
    399, 162, 758, 219, 918, 237, 412, 566, 826, 248, 866, 950, 626, 949, 687, 217,
    815, 67, 104, 58, 512, 24, 892, 894, 767, 553, 81, 379, 843, 831, 445, 742, 717,
    958,743, 527
]

# Iterate through the numbers in the list.
for x in numbers:
  if x == 237:
      print(x) # Print the number if its 237
      break # Exit the loop if 237 is found.
  elif x % 2 == 0:
      print(x) # Print the number if its even.

386
462
418
344
236
566
978
328
162
758
918
237


**Explanation:**
The said code defines a list called "numbers" with a set of integers. Then it uses a for loop to iterate through each element in the "numbers" list.
For each iteration, it checks if the current element is equal to 237, if true it prints the current element and exit the loop using the "break" statement.
If the current element is not equal to 237, it checks if the current element is even, if true it prints the current element.
Using this logic, the code will only print the number 237 and even numbers of the list. If 237 was not found in the list, the code will print all the even numbers in the list.

## #29
Write a Python program that prints out all colors from color_list_1 that are not present in color_list_2.

In [55]:
# Create two sets, color_list_1 and color_list_2.
color_list_1 = set(["White", "BlacK", "Red"])
color_list_2 = set(["Red", "Green"])

# Print the original elements of the sets.
print("Original set elements:")
print(color_list_1)
print(color_list_2)

# Calculate and print the difference of color_list_1 and color_list_2.
print("\nDifference of color_list_1 and color_list_2:")
print(color_list_1.difference(color_list_2))

# Calculate and print the difference of color_list_2 and color_list_1.
print("\nDifference of color_list_2 and color_list_1:")
print(color_list_2.difference(color_list_1))

Original set elements:
{'White', 'Red', 'BlacK'}
{'Green', 'Red'}

Difference of color_list_1 and color_list_2:
{'White', 'BlacK'}

Difference of color_list_2 and color_list_1:
{'Green'}


**Explanation:**
The said code defines two sets called "color_list_1" and "color_list_2" containing string elements.
The elements of both sets are then printed.
After that, it uses the difference() method on "color_list_1" and passes "color_list_2" as an argument to it. This method returns the elements that are in the "color_list_1" but not in "color_list_2".
Again it uses the difference() method on "color_list_2" and passes "color_list_1" as an argument to it. This method returns the elements that are in the "color_list_2" but not in "color_list_1".
The output of the code will be:
Original set elements:
{'Red', 'White', 'Black'}
{'Red', 'Green'}
Differenct of color_list_1 and color_list_2:
{'White', 'Black'}
Differenct of color_list_2 and color_list_1:
{'Green'}

## #30
Write a Python program that will accept the base and height of a triangle and compute its area.

In [56]:
# Prompt the user to input the base and height of a triangle as integers.
b = int(input("Input the base: "))
h = int(input("Input the height: "))

# Calculate the area of the triangle using the formula: (base * height) / 2
area = b * h / 2

# Print the calculated area of the triangle.
print("area = ", area)

Input the base: 54
Input the height: 18
area =  486.0


**Explanation:**
The said code prompts the user to input an integer value for the base and height of a triangle. Stores these values in variables "b" and "h" respectively. It then calculates the area of the triangle by multiplying the base and height and dividing the result by 2 and assigns the result to the variable "area". Finally, it prints the calculated area.

## #31
Write a Python program that computes the greatest common divisor (GCD) of two positive integers

In [57]:
# Define a function to calculate the greatest common divisor (GCD) of two numbers.
def gcd(x, y):
  # Initialize gcd to 1.
  gcd = 1

  # Check if y is a divisor of x (x is divisible by y)
  if x % y == 0:
    return y

  # Iterate from half of y down to 1.
  for k in range(int(y / 2), 0, -1):
    # Check if both x and y are divisible by k.
    if x % k == 0 and y % k == 0:
      # Update the GCD to the current value of k and exit the loop.
      gcd = k
      break

  # Return the calculated GCD
  return gcd

# Print the GCD of specific pairs of numbers.
print("GCD of 12 & 17 =", gcd(12, 17))
print("GCD of 4 & 6 =", gcd(4, 6))
print("GCD of 336 & 360 =", gcd(336, 360))

GCD of 12 & 17 = 1
GCD of 4 & 6 = 2
GCD of 336 & 360 = 24


**Explanation:**

Here is a simple implementation of the Euclidean algorithm for finding the greatest common divisor (GCD) of two integers.

The function "gcd(x,y)" starts by initializing the variable "gcd" to 1, and then checks if "x" is divisible by "y". If it is, it returns "y" as the GCD.

If "x" is not divisible by "y", it enters a "for" loop. In each iteration, it checks if "k" is a common divisor of "x" and "y", and if it is, it assigns the value of k to the variable "gcd" and breaks out of the loop. Finally, the function returns the value of GCD of "x" and "y".

Last three print statements of the said code call the function with different inputs, and print the output as "GCD of 12 & 17 = 1", "GCD of 4 & 6 = 2" and "GCD of 336 & 360 = 24".

## #32
Write a Python program to find the least common multiple (LCM) of two positive integers

In [58]:
# Define a function 'lcm' that calculates the least common multiple (LCM) of two numbers, 'x' and 'y'.
def lcm(x, y):
  # Compare 'x' and 'y' to determine the larger number and store it in 'z'.
  if x > y:
    z = x
  else:
    z = y

  # Use a 'while' loop to find the LCM.
  while True:
    # Check if 'z' is divisible by both 'x' and 'y' with no remainder.
    if (z % x == 0) and (z % y == 0):
      # If both conditions are met, 'z' is the LCM, so store it in 'lcm' and break the loop.
      lcm = z
      break
    # If the conditions are not met, increment 'z' and continue checking.
    z += 1

  # Return the calculated LCM.
  return lcm

# Calculate and print the LCM of 4 and 6
print(lcm(4, 6))
# Calculate and print the LCM of 15 and 17.
print(lcm(15, 17))

12
255


**Explanation:**

The said code defines a function "lcm(x, y)" that takes two integers as its argument and will return the least common multiple (LCM) of two arguments, "x" and "y".

First, the function determines which number is larger and assigns that value to a variable named "z". It then enters a loop that continues indefinitely (while(True)) until the LCM is found.

Inside the loop, the code checks if the current value of "z" is divisible by both "x" and "y" (if((z % x == 0) and (z % y == 0))). If it is true then the LCM is assigned to the variable "lcm" and the loop is broken. Otherwise, the value of "z" is incremented by 1 and the check is repeated.

Finally, the function returns the value of "lcm" as the output.

After the function definition, the code calls the function twice, first with the arguments 4 and 6, then with the arguments 15 and 17 and prints the results.

## #33
Write a Python program to sum three given integers. However, if two values are equal, the sum will be zero.

In [59]:
# Define a function 'sum_three' that takes three integer inputs: x, y, and z.
def sum_three(x, y, z):
  # Check if any of the two input calues are equal. If so, set 'sum' to 0.
  if x == y or y == x or x == z:
    sum = 0
  else:
    # If all three input values are distinct, calculate the sum of x, y, and z.
    sum = x + y + z
  # Return the calculated sum.
  return sum

# Test the 'sum_three' function with different sets of input values and print the results.
print(sum_three(2, 1, 2))
print(sum_three(3, 2, 2))
print(sum_three(2, 2, 2))
print(sum_three(1, 2, 3))

0
7
0
6


**Explanation:**

The said code defines a function "sum_three(x, y, z)" that takes three integers as its argument and returns their sum.

The function first checks if any of the three integers (x,y,z) are equal, if any two or three integers are equal, it assigns 0 to the variable "sum".

Otherwise, it assigns the sum of the three integers to the variable "sum".

Finally, it returns the value of "sum".

At the end, the function is called four times with different inputs and prints the sum of the inputs.


## #34
Write a Python program to sum two given integers. However if the sum is between 15 and 20 it will return 20.

In [60]:
# Define a function 'sum' that takes two integer inputs: x and y
def sum(x, y):
  # Calculate the sum of x and y and store it in the 'sum' variable
  sum = x + y
  # Check if the calculated sum is within the range [15, 20) (inclusive on 15, exclusive on 20).
  if sum in range(15, 20):
    # If the sum is within the range, return 20.
    return 20
  else:
    # If the sum is outside the range, return the calculated sum.
    return sum

# Test the 'sum' function with different sets of input values and print the results
print(sum(10, 6))
print(sum(10, 2))
print(sum(10, 12))

20
12
22


**Explanation:**

The said code defines a function "sum(x, y)" that takes two integers as arguments and returns their sum.

It first calculates the sum of the two integers and assigns it to the variable "sum".

Then it checks whether the value of "sum" is in the range of 15 to 20 (inclusive) using the in-built Python function "range()".

If the value of "sum" is in the range of 15 to 20, the function returns 20 otherwise, it returns the value of "sum".

At the end, the function is called three times with different inputs and prints the sum of the inputs.

## #35
Write a Python program that returns true if the two given integer values are equal or their sum or difference is 5.

In [61]:
# Define a function 'test_number5' that takes two integer inputs: x and y.
def test_number5(x, y):
  # Check if any of the following conditions are met:
  # 1. x is equal to y.
  # 2. The absolute difference between x and y is equal to 5
  # 3. The sum of x and y is equal to 5.
  if x == y or abs(x - y) == 5 or (x + y) == 5:
    # If any of the conditions are met, return True.
    return True
  else:
    # If none of the conditions are met, return False.
    return False

# Test the 'test_number5' function with different sets of input values and print the results.
print(test_number5(7, 2))
print(test_number5(3, 2))
print(test_number5(2, 2))
print(test_number5(7, 3))
print(test_number5(27, 53))

True
True
True
False
False


**Explanation:**

The said code defines a function "test_number5(x, y)" that takes two integers "x" and "y" as arguments and check if they are equal, or if their sum or difference is equal to 5. If any of these conditions is true, the program would return "True", otherwise, it would return "False".

At the end, the function is called five times with different inputs and prints the outputs of the inputs.

## #36